### ISSUE TITLE
Docker Hash Lookup v3

### ISSUE THREAD

**Issue by Horneth** [broadinstitute/cromwell#2048]
This proposal should give us more flexibility regarding docker tags while keeping the call caching safety on false positive / negative.

Docker runtime attributes with docker hashes do not need any additional processing. All logic in this ticket applies to docker runtime attributes with a "floating" tag, which will be referred as "tag" in this issue.

In all cases, if Cromwell fails to retrieve the docker hash for a task, for any reason, the corresponding call(s) will NOT be eligible for call caching, neither read nor write, regardless of the call caching configuration in effect.

**When to get the hashes and what to do with them:**

1. Cromwell will lookup the hashes corresponding to docker tags, for all docker attributes in all tasks in a workflow and its subworkflows, at Materialization time.
If the runtime attribute value can't be determined, the task in question will be ineligible for call caching. The only case when that should be true is if the attribute is an expression with variables depending on previous tasks being run.

2. If the hash lookup succeed, Cromwell will use that hash to perform any call cache read / write according to the call caching configuration in effect. It will also provide that hash, along with the original floating tag, to the backend when the job gets dispatched.

3. Backends will choose wether to use the hash or the floating tag. They will report to the engine which one they used, so that the engine can send this information to the metadata.

**How to get the hash:**

1. How to get the hash depends on the backend. Which means, at this time, that only workflows for which the backend is known statically at workflow submission time will be supported.

2. If the task is expected to run on the **Local Backend**, Cromwell will attempt to find the hash corresponding to the tag on the machine where it's being run. This first attempt must be done without executing a `pull` to avoid overriding the current local image, if it exits, with the remote repository version.
If the image is not present locally, cromwell will attempt to `pull` the image locally, and use the hash from the newly retrieved image.

3. If the task is expected to run on a **Non Local Backend**, cromwell will attempt to retrieve the image from a remote repository. The `DockerHashActor` can be used for that effect.

**Comment by cjllanwarne** [comment]
Looks awesome to me. Thinking out loud.....:

 - If a hash lookup fails (i.e. I mean the dockerhub request, not the expression evaluation), not even starting the workflow might be the expected behaviour, because "don't waste my money starting an entire analysis when in 2 minutes I can resubmit and get CC"?
 - Should we also have a "disable docker hash cache" option, for Lee's very-fast iterations? Or just a "clear hash cache" REST endpoint?
 - The hash lookup cache may also need to be aware of local vs remote hashes
 - ToL: one step further away from dynamic backend assignment 😢

**Comment by Horneth** [comment]
For the second point, it's already possible, setting the cache-size to 0 will disable the cache here:
https://github.com/broadinstitute/cromwell/blob/develop/core/src/main/resources/reference.conf#L150

**Comment by mcovarr** [comment]
Looking at the existing `JobPreparationActor` code I saw that the Docker hash credentials are being created by calling this method in `BackendLifecycleActorFactory`:
```
def dockerHashCredentials(initializationDataOption: Option[BackendInitializationData]): List[Any] = List.empty
```
The JES backend overrides this to return a non-empty `List`.  Since we don't yet have the required `BackendInitializationData` during workflow materialization, @Horneth suggested the Docker hash calculation be performed slightly downstream in workflow initialization instead.

**Comment by mcovarr** [comment]
Thoughts for a Monday Tech Talk™️:

Say we run a workflow with a 100-wide scatter and the floating Docker tags are resolved to hashes during workflow initialization.  90 of the jobs launch, but then the server goes down.  The server comes up some time later and recovers the first 90 running jobs, but in the meantime the floating Docker tag has moved to a new version.  We rerun the workflow initialization and calculate a different hash for the remaining 10 shards of the scatter.  

It seems the hash for a Docker tag for a particular workflow should be persisted to be able to handle this case.  I also wonder per @cjllanwarne if we shouldn't keep this activity in the `JobPreparationActor` to avoid knowingly creating a system that we'll have to replace when we implement dynamic dispatch.  Keeping this in `JobPreparationActor` would also give us greater ability to resolve expressions for Docker tags than if we do this earlier in the workflow lifecycle.

**Comment by mcovarr** [comment]
It looks like everything here has been addressed in #2282 so I think this should be closeable once that merges.
